### ISSUE TITLE
Real explanation of entryPointStrategy

### ISSUE THREAD

**Issue by 6nv** [TypeStrong/typedoc#2487]
## Search terms

Entry point strategy, entryPointStrategy, expand, resolve, not included in documentation

## Environment

TypeDoc 0.25.7
Using TypeScript 4.9.5

## Question

Hi all, I've been experimenting with TypeDoc, and it's worked great so far. However, [EntryPointStrategy](https://typedoc.org/api/enums/EntryPointStrategy.html) is very confusing.

Assume my file structure is:

```
tsconfig.json
package.json
...
src/
  index.ts
  other.ts
```

When I run `typedoc src` I get this warning:

```
[warning] OtherClass, defined in ./src/other.ts, is referenced by IndexClass.a but not included in the documentation.
```

Now when I run `typedoc src/*` _OR_ `typedoc --entryPointStrategy expand src`, it compiles.

But, what if I change my file structure to this:

```
tsconfig.json
package.json
...
src/
  index.ts
  sub/
    other.ts
```

Expectedly, `typedoc src/*` returns another warning.

```
[warning] The entry point ./src/sub is not included in the program for your provided tsconfig. If you wanted to include files inside this directory, set --entryPointStrategy to expand or specify a glob.
[warning] OtherClass, defined in ./src/sub/other.ts, is referenced by index.IndexClass.a but not included in the documentation.
```

But, `typedoc --entryPointStrategy expand src` still works.

So, to me it seems like you would want to use `--entryPointStrategy expand` every time, but according to the documentation, the default strategy was changed in v0.21.

1. Why was it changed?
2. What really _is_ the difference between these entry point strategies?
3. What is the best practice moving forward to ensure all documentation is generated?

**Comment by Gerrit0** [comment]
1. It was changed in an attempt to make entry points for documentation by default closer to what you'd get if you'd done `require("./path/to/entry")`. Generally, libraries will re-export everything from an `index.ts` file so that users don't have to import from `lib/dist/some/path/to/what/is/actually/wanted`.

2. The difference is exactly what it says on that page you linked:
    > Behaves like the resolve behavior, but will recursively expand directories into an entry point for each file within the directory.

    In practice, the difference is just one extra function call for expand in TypeDoc. The expand option takes every path matched by the entry point glob and, if it is a directory, searches for files within it.

3. It depends, how do you expect your library to be used?
    If you expect users to import from `lib/sub/other` or any other path in your project, then it makes sense for that to be an entry point, and `--entryPointStrategy expand` is probably what you want. You could also use `src/**/*.ts` with either strategy. (Less ideal if you have `mts` or `tsx` files)

    If you expect users to only import from `lib`, then you should just create documentation for `src/index.ts`, TypeDoc's [treatValidationWarningsAsErrors](https://typedoc.org/options/validation/#treatvalidationwarningsaserrors) option can be used to fail the build if you forgot to export something that's referenced

**Comment by 6nv** [comment]
> Generally, libraries will re-export everything from an index.ts file so that users don't have to import from lib/dist/some/path/to/what/is/actually/wanted.

Awesome! I didn't realize I needed to re-export in index.ts. I tried this and it is spectacular. Doing it that way makes it much nicer and no need for the entry point strategy (so I am using the default value of `resolve` here.)

And thanks for pointing out the command flag. I want to enforce good code practice on myself so these will be good to check out.
